File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.js
function solution(user_id, banned_id) {
    const isMatch = (id, ban) => {
        if(id.length !== ban.length) return false;
        for(let i = 0; i < id.length; i++) {
            if(ban[i] !== '*' && ban[i] !== id[i]) return false;
        }
        return true;
    }
    
    const possibleSet = new Set();
    const findCombinations = (currentIndex, selectedUsers) => {
        if(currentIndex === banned_id.length) {
            possibleSet.add([...selectedUsers].sort().join(',')); // 중복 방지
            return;
        }
        user_id.forEach(user => {
            if(!selectedUsers.includes(user) && isMatch(user, banned_id[currentIndex])) {
                selectedUsers.push(user);
                findCombinations(currentIndex + 1, selectedUsers);
                selectedUsers.pop(user);
            }
        })
    }
    findCombinations(0, []);
    
    return possibleSet.size;
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
def solution(user_id, banned_id):
    answer_set = set()
    
    def is_matched(user, ban):
        if len(user) != len(ban):
            return False
        for i in range(len(user)):
            if ban[i] != '*' and user[i] != ban[i]:
                return False
        return True
    
    def find_combis(cur_idx, selected_users):
        if cur_idx == len(banned_id):
            answer_set.add(",".join(sorted(selected_users)))
            return
        
        for user in user_id:
            if user not in selected_users and is_matched(user, banned_id[cur_idx]):
                selected_users.append(user)
                find_combis(cur_idx + 1, selected_users)
                selected_users.pop()
    
    find_combis(0, [])
    return len(answer_set)
        
        